docs/proposals/2022-07-13-appset-progressive-rollout-strategy.md

```diff
@@ -0,0 +1,165 @@
+Steps for the rolling update are defined by a list of matchExpression label selectors. Each step must finish updating before the next step advances. If steps are left undefined the application update order is deterministic.
+
+Complete ApplicationSet spec example.
+```
+apiVersion: argoproj.io/v1alpha1
+kind: ApplicationSet
+metadata:
+  name: guestbook
+spec:
+  generators:
+  - list:
+      elements:
+      - cluster: engineering-dev
+        url: https://1.2.3.4
+      - cluster: engineering-prod
+        url: https://2.4.6.8
+      - cluster: finance-preprod
+        url: https://9.8.7.6/
+  strategy:
+    type: RollingUpdate
+    rollingUpdate:
+      steps:
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - dev
+          maxUpdate: 0 # if undefined or 0, all applications matched are updated together
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - qa
+        - matchExpressions:
+            - key: env
+              operator: In
+              values:
+                - us-east-2
+                - eu-west-1
+                - ap-southeast-1
+          maxUpdate: 1 # maxUpdate supports both integer and percentage string values
+  template:
+    metadata:
+      name: '{{cluster}}-guestbook'
+    spec:
+      source:
+        repoURL: https://github.com/infra-team/cluster-deployments.git
+        targetRevision: HEAD
+        path: guestbook/{{cluster}}
+      destination:
+        server: '{{url}}'
+        namespace: guestbook
+```
+
+In the above example, when the guestbook ApplicationSet is created or modified, the Application resources are each updated in the order defined in `strategy.rollingUpdate`. In this case, all Applications whose desired resource state matches the label expression `env: dev` are updated to match the template, as the `maxUpdate` is set to zero. The rolling update strategy progresses after the first set of Applications has successfully completed. Progress towards the next step starts only after the current step has completely finished, regardless of the `maxUpdate` value. The `maxUpdate` field only throttles the total number of matching Applications updating in the current step. After the first step completes, the ApplicationSet updates all Application resources with label `env: qa` at the same time, because `maxUpdate` is undefined. Finally, during the third step, the Application resources labeled `env: us-east-2`, `env: eu-west-1`, or `env: ap-southeast-1` are updated, one by one, as the `maxUpdate` for the final step is 1.
+
+An Application rollout is considered “complete” when the Application resource has been:
+- Synced successfully.
+- Moved into a “Progressing” state.
+- Moved out of a “Progressing” state and into a “Healthy” state.
```

Comment:
Is checking each of these necessary, or would it be sufficient to just check for a Healthy state? (I'm not sure exactly in what order the Application controller does things and whether an App _must_ enter a Progressing state after a sync).

Reply:
The problem I ran into while implementing this is that the ApplicationSet controller and the Application controllers work entirely independently. Because of that, there's a period of time where an Application resource has been updated by the ApplicationSet controller, but has not yet been put into the "Progressing" state by the Application controller.

If the ApplicationSet controller is only looking for a "Healthy" state after updating the Application resource, it will see the lingering "Healthy" state and move on to the next rollout step prematurely.

Reply:
Yep, that makes sense. Leads to my question about handling cases where not all Applications actually require an update. Will look at your response for that one.  :-)